File: src/main/java/net/azisaba/townia/data/Town.java
package net.azisaba.townia.data;

import java.util.UUID;

public class Town {

    private final UUID id;
    private String name;
    private UUID mayorUuid;
    private UUID nationUuid;
    private double balance;
    private int claimLimit;
    private int bonusClaims;
    private boolean isPublic;
    private long createdAt;

    // Extended features
    private String board;
    private double taxes;
    private double plotPrice;
    private boolean pvp;
    private boolean mobs;
    private boolean explosions;
    private boolean fire;
    
    // Home block
    private String homeBlockWorld;
    private int homeBlockX;
    private int homeBlockZ;
    
    // Permissions
    private String permsResident = "BDSI";
    private String permsAlly = "";
    private String permsOutsider = "";
    private String permsNation = "";
    
    private double dailyUpkeep;

    // Spawn point
    private String spawnWorld;
    private double spawnX;
    private double spawnY;
    private double spawnZ;
    private float spawnYaw;
    private float spawnPitch;

    public Town(UUID id, String name, UUID mayorUuid, UUID nationUuid,
                double balance, int claimLimit, int bonusClaims, boolean isPublic,
                long createdAt, String board, double taxes, double plotPrice,
                boolean pvp, boolean mobs, boolean explosions, boolean fire,
                String spawnWorld, double spawnX, double spawnY, double spawnZ,
                float spawnYaw, float spawnPitch) {
        this.id = id;
        this.name = name;
        this.mayorUuid = mayorUuid;
        this.nationUuid = nationUuid;
        this.balance = balance;
        this.claimLimit = claimLimit;
        this.bonusClaims = bonusClaims;
        this.isPublic = isPublic;
        this.createdAt = createdAt;
        this.board = board;
        this.taxes = taxes;
        this.plotPrice = plotPrice;
        this.pvp = pvp;
        this.mobs = mobs;
        this.explosions = explosions;
        this.fire = fire;
        this.spawnWorld = spawnWorld;
        this.spawnX = spawnX;
        this.spawnY = spawnY;
        this.spawnZ = spawnZ;
        this.spawnYaw = spawnYaw;
        this.spawnPitch = spawnPitch;
    }

    public UUID getId() { return id; }
    public String getName() { return name; }
    public UUID getMayorUuid() { return mayorUuid; }
    public UUID getNationUuid() { return nationUuid; }
    public double getBalance() { return balance; }
    public int getClaimLimit() { return claimLimit; }
    public int getBonusClaims() { return bonusClaims; }
    public int getTotalClaimLimit() { return claimLimit + bonusClaims; }
    public boolean isPublic() { return isPublic; }
    public long getCreatedAt() { return createdAt; }
    
    public String getBoard() { return board; }
    public double getTaxes() { return taxes; }
    public double getPlotPrice() { return plotPrice; }
    public boolean hasPvp() { return pvp; }
    public boolean hasMobs() { return mobs; }
    public boolean hasExplosions() { return explosions; }
    public boolean hasFire() { return fire; }

    public String getSpawnWorld() { return spawnWorld; }
    public double getSpawnX() { return spawnX; }
    public double getSpawnY() { return spawnY; }
    public double getSpawnZ() { return spawnZ; }
    public float getSpawnYaw() { return spawnYaw; }
    public float getSpawnPitch() { return spawnPitch; }
    public boolean isInNation() { return nationUuid != null; }
    public boolean hasSpawn() { return spawnWorld != null; }
    
    public String getHomeBlockWorld() { return homeBlockWorld; }
    public int getHomeBlockX() { return homeBlockX; }
    public int getHomeBlockZ() { return homeBlockZ; }
    public boolean hasHomeBlock() { return homeBlockWorld != null; }
    
    public String getPermsResident() { return permsResident; }
    public String getPermsAlly() { return permsAlly; }
    public String getPermsOutsider() { return permsOutsider; }
    public String getPermsNation() { return permsNation; }
    
    public double getDailyUpkeep() { return dailyUpkeep; }

    public void setName(String name) { this.name = name; }
    public void setMayorUuid(UUID mayorUuid) { this.mayorUuid = mayorUuid; }
    public void setNationUuid(UUID nationUuid) { this.nationUuid = nationUuid; }
    public void setBalance(double balance) { this.balance = balance; }
    public void setClaimLimit(int claimLimit) { this.claimLimit = claimLimit; }
    public void setBonusClaims(int bonusClaims) { this.bonusClaims = bonusClaims; }
    public void setPublic(boolean aPublic) { isPublic = aPublic; }

    public void setBoard(String board) { this.board = board; }
    public void setTaxes(double taxes) { this.taxes = taxes; }
    public void setPlotPrice(double plotPrice) { this.plotPrice = plotPrice; }
    public void setPvp(boolean pvp) { this.pvp = pvp; }
    public void setMobs(boolean mobs) { this.mobs = mobs; }
    public void setExplosions(boolean explosions) { this.explosions = explosions; }
    public void setFire(boolean fire) { this.fire = fire; }

    public void setSpawn(String world, double x, double y, double z, float yaw, float pitch) {
        this.spawnWorld = world;
        this.spawnX = x;
        this.spawnY = y;
        this.spawnZ = z;
        this.spawnYaw = yaw;
        this.spawnPitch = pitch;
    }
    
    public void setHomeBlock(String world, int x, int z) {
        this.homeBlockWorld = world;
        this.homeBlockX = x;
        this.homeBlockZ = z;
    }
    
    public void setPermsResident(String perms) { this.permsResident = perms; }
    public void setPermsAlly(String perms) { this.permsAlly = perms; }
    public void setPermsOutsider(String perms) { this.permsOutsider = perms; }
    public void setPermsNation(String perms) { this.permsNation = perms; }
    
    public void setDailyUpkeep(double dailyUpkeep) { this.dailyUpkeep = dailyUpkeep; }
    
    // Additional constructor for loading
    public Town(UUID id, String name, UUID mayorUuid, UUID nationUuid,
                double balance, int claimLimit, int bonusClaims, boolean isPublic,
                long createdAt, String board, double taxes, double plotPrice,
                boolean pvp, boolean mobs, boolean explosions, boolean fire,
                String spawnWorld, double spawnX, double spawnY, double spawnZ,
                float spawnYaw, float spawnPitch,
                String homeBlockWorld, int homeBlockX, int homeBlockZ,
                String permsResident, String permsAlly, String permsOutsider, String permsNation,
                double dailyUpkeep) {
        this(id, name, mayorUuid, nationUuid, balance, claimLimit, bonusClaims, isPublic, createdAt,
             board, taxes, plotPrice, pvp, mobs, explosions, fire, spawnWorld, spawnX, spawnY, spawnZ, spawnYaw, spawnPitch);
        this.homeBlockWorld = homeBlockWorld;
        this.homeBlockX = homeBlockX;
        this.homeBlockZ = homeBlockZ;
        this.permsResident = permsResident;
        this.permsAlly = permsAlly;
        this.permsOutsider = permsOutsider;
        this.permsNation = permsNation;
        this.dailyUpkeep = dailyUpkeep;
    }
}

File: src/main/java/net/azisaba/townia/data/Town.kt
package net.azisaba.townia.data

import net.azisaba.townia.Townia
import java.util.*

class Town(
    val id: UUID?, var name: String?, var mayorUuid: UUID?, var nationUuid: UUID?,
    var balance: Double, var claimLimit: Int, var bonusClaims: Int, var isPublic: Boolean,
    val createdAt: Long,
    var board: String?, var taxes: Double, var plotPrice: Double,
    private var pvp: Boolean, private var mobs: Boolean, private var explosions: Boolean, private var fire: Boolean,
    // Spawn point
    var spawnWorld: String?, var spawnX: Double, var spawnY: Double, var spawnZ: Double,
    var spawnYaw: Float, var spawnPitch: Float
) {
    var isOpen: Boolean = false

    // Home block
    var homeBlockWorld: String? = null
        private set
    var homeBlockX: Int = 0
        private set
    var homeBlockZ: Int = 0
        private set

    var permsResident: String? = "BDSI"
    var permsAlly: String? = ""
    var permsOutsider: String? = ""
    var permsNation: String? = ""

    var dailyUpkeep: Double = 0.0

    var tag: String? = null
    var isTaxPercent: Boolean = false
    var plotTax: Double = 0.0
    var shopPrice: Double = 0.0
    var shopTax: Double = 0.0
    var embassyPrice: Double = 0.0
    var embassyTax: Double = 0.0

    var isAllowInvisibility: Boolean = true
    var isAllowSit: Boolean = true
    var isAllowPetPickup: Boolean = true
    var isAllowPassenger: Boolean = true


    val outposts: MutableList<TowniaOutpost?> = ArrayList<TowniaOutpost?>()

    val totalClaimLimit: Int
        get() {
            val resCount = Townia.instance.residentManager.getResidentsByTown(id!!).size
            val townLevel = Townia.instance.towniaConfig.getTownLevel(resCount)
            val baseLimit = townLevel?.townBlockLimit ?: claimLimit
            
            var nationBonus = 0
            if (isInNation) {
                val nationOpt = Townia.instance.nationManager.getNation(nationUuid!!)
                if (nationOpt.isPresent) {
                    nationBonus = nationOpt.get().getTownBlockLimitBonus()
                }
            }
            
            return baseLimit + bonusClaims + nationBonus
        }

    fun getFormattedName(sender: org.bukkit.command.CommandSender? = null): String {
        val resCount = Townia.instance.residentManager.getResidentsByTown(id!!).size
        val townLevel = Townia.instance.towniaConfig.getTownLevel(resCount)
        val prefix = townLevel?.namePrefix ?: ""
        val postfix = townLevel?.namePostfix ?: ""
        return "${Townia.instance.messageManager.processTranslationKeys(prefix, sender)}${name ?: ""}${Townia.instance.messageManager.processTranslationKeys(postfix, sender)}"
    }

    fun getMayorPrefix(sender: org.bukkit.command.CommandSender? = null): String {
        val resCount = Townia.instance.residentManager.getResidentsByTown(id!!).size
        val townLevel = Townia.instance.towniaConfig.getTownLevel(resCount)
        val prefix = townLevel?.mayorPrefix ?: ""
        return Townia.instance.messageManager.processTranslationKeys(prefix, sender)
    }

    fun getMayorPostfix(sender: org.bukkit.command.CommandSender? = null): String {
        val resCount = Townia.instance.residentManager.getResidentsByTown(id!!).size
        val townLevel = Townia.instance.towniaConfig.getTownLevel(resCount)
        val postfix = townLevel?.mayorPostfix ?: ""
        return Townia.instance.messageManager.processTranslationKeys(postfix, sender)
    }

    fun hasPvp(): Boolean {
        return pvp
    }

    fun hasMobs(): Boolean {
        return mobs
    }

    fun hasExplosions(): Boolean {
        return explosions
    }

    fun hasFire(): Boolean {
        return fire
    }

    val isInNation: Boolean
        get() = nationUuid != null

    fun hasSpawn(): Boolean {
        return spawnWorld != null
    }

    fun hasHomeBlock(): Boolean {
        return homeBlockWorld != null
    }

    val townSizeRank: String
        get() {
            val count: Int = Townia.instance.residentManager.getResidentsByTown(id!!).size
            if (count >= 28) return "MetroPolis"
            if (count >= 24) return "Large City"
            if (count >= 20) return "City"
            if (count >= 14) return "Large Town"
            if (count >= 10) return "Town"
            if (count >= 6) return "Village"
            if (count >= 2) return "Hamlet"
            return "Settlement"
        }

    fun setPvp(pvp: Boolean) {
        this.pvp = pvp
    }

    fun setMobs(mobs: Boolean) {
        this.mobs = mobs
    }

    fun setExplosions(explosions: Boolean) {
        this.explosions = explosions
    }

    fun setFire(fire: Boolean) {
        this.fire = fire
    }

    fun setSpawn(world: String?, x: Double, y: Double, z: Double, yaw: Float, pitch: Float) {
        this.spawnWorld = world
        this.spawnX = x
        this.spawnY = y
        this.spawnZ = z
        this.spawnYaw = yaw
        this.spawnPitch = pitch
    }

    fun setHomeBlock(world: String?, x: Int, z: Int) {
        this.homeBlockWorld = world
        this.homeBlockX = x
        this.homeBlockZ = z
    }

    constructor(
        id: UUID?, name: String?, mayorUuid: UUID?, nationUuid: UUID?,
        balance: Double, claimLimit: Int, bonusClaims: Int, isPublic: Boolean,
        createdAt: Long, board: String?, taxes: Double, plotPrice: Double,
        pvp: Boolean, mobs: Boolean, explosions: Boolean, fire: Boolean,
        spawnWorld: String?, spawnX: Double, spawnY: Double, spawnZ: Double,
        spawnYaw: Float, spawnPitch: Float,
        homeBlockWorld: String?, homeBlockX: Int, homeBlockZ: Int,
        permsResident: String?, permsAlly: String?, permsOutsider: String?, permsNation: String?,
        dailyUpkeep: Double
    ) : this(
        id, name, mayorUuid, nationUuid, balance, claimLimit, bonusClaims, isPublic, createdAt,
        board, taxes, plotPrice, pvp, mobs, explosions, fire, spawnWorld, spawnX, spawnY, spawnZ, spawnYaw, spawnPitch
    ) {
        this.homeBlockWorld = homeBlockWorld
        this.homeBlockX = homeBlockX
        this.homeBlockZ = homeBlockZ
        this.permsResident = permsResident
        this.permsAlly = permsAlly
        this.permsOutsider = permsOutsider
        this.permsNation = permsNation
        this.dailyUpkeep = dailyUpkeep
    }
}